dtest_setup.py

```diff
@@ -350,17 +350,35 @@ def stop_active_log_watch(self):
         """
         self.log_watch_thread.join(timeout=60)
 
+    def stop_cluster(self, gently=False):
+        """
+        Stops the cluster; if 'gently' is requested and a NodeError occurs, then
+        try again without 'gently'.
+
+        Some tests, by design, leave the cluster in a state which prevents it from
+        being stopped using 'gently'. Retrying without 'gently' will avoid marking
+        the test as a failure, but may prevent jacoco results from being recorded.
+        """
+        try:
+            self.cluster.stop(gently)
+        except NodeError as e:
+            if gently:
+                logger.debug("Exception stopping cluster with gently=True, retrying with gently=False: {0}".format(e))
+                self.cluster.stop(gently=False)
+            else:
+                raise e
+
     def cleanup_cluster(self, request=None):
         with log_filter('cassandra'):  # quiet noise from driver when nodes start going down
             test_failed = request and hasattr(request.node, 'rep_call') and request.node.rep_call.failed
             if self.dtest_config.keep_test_dir or (self.dtest_config.keep_failed_test_dir and test_failed):
-                self.cluster.stop(gently=self.dtest_config.enable_jacoco_code_coverage)
+                self.stop_cluster(gently=self.dtest_config.enable_jacoco_code_coverage)
```

Comment:
Note: looking at this logic - unless I am not fully awake -  I think it can be simplified:
```
self.stop_cluster(gently=self.dtest_config.enable_jacoco_code_coverage)
preserve_logs = self.dtest_config.keep_test_dir or (self.dtest_config.keep_failed_test_dir and test_failed)
if not preserve_logs:
    # Cleanup everything
```
but I understand we do not want to introduce changes vs apache
```

Reply:
> After this change if for any reason some other tests will stop closing gently we will not notice it as test failures, potentially missing a bug.

Actually, I don't think that would happen. The only time that stop(gently) is for jacoco code coverage; in all other cases it is stop(kill). If stop(gently) was the normal case, this might be true, but stop(kill) is the norm.

> I would rather explicitly stop by kill in this particular test case at the end of the test.
> Assuming that C* devs confirm that this is expected behavior after STOP policy is executed.
> 
> Ideal approach would be to add some decorator of `cluster` flag so that `ccm` itself knows that in this case it should not try to stop gently (and save some test duration).

To save some test duration by avoiding a stop(gently) that is expected to fail, maybe we can annotate this test (and others) so that cleanup_cluster can use that to override the gentle request for jacoco.

I suspected there are other tests besides `test_stop_failure_policy` and `test_stop_commit_failure_policy` failing for the same reason with jacoco and it seemed the current approach has the advantage of working for those as well. But maybe a mechanism to tag such tests is better in making the expectation clear.

Reply:
> Note: looking at this logic - unless I am not fully awake - I think it can be simplified:

I noticed it too, and agree it can be simplified to be more easily read/understood.

> but I understand we do not want to introduce changes vs apache

I am always debating this question with myself; in this case, I am wondering if this issue is affecting OSS also and therefore the fix should be ported back? In that case, I would be in favor of making this logic simpler.